w7/assignment7_4_1/animals.cpp

```diff
@@ -0,0 +1,151 @@
+#include <iostream>
+#include <vector>
+
+// Base class
+class Animal
+{
+public:
+    // Default constructor.
+    Animal(){};
+    // Constructor with parameter
+    Animal(std::string name)
+    {
+        name_ = name;
+    }
+    // Pure virtual function
+    virtual void make_sound() const = 0;
+    // Abstract but not pure virtual function.
+    virtual void hair() const {};
+    virtual std::string get_name() const
+    {
+        return name_;
+    }
+    virtual ~Animal() = default;
+
+private:
+    std::string name_{};
+};
+
+/* Lizard inherits Animal
+*/
+class Lizard : public Animal
+{
+public:
+    Lizard()
+    {
+        name_ = "Leevi";
+    };
+    Lizard(std::string name)
+    {
+        name_ = name;
+    }
+
+    void make_sound() const override
+    {
+        std::cout << "Zzzsssh\n";
+    }
+
+    std::string get_name() const override
+    {
+        return name_;
+    }
+
+private:
+    std::string name_{};
+};
+
+/* Dog inherits Animal
+*/
+class Dog : public Animal
+{
+public:
+    Dog()
+    {
+        name_ = "Tessu";
+    };
+    Dog(std::string name)
+    {
+        name_ = name;
+    }
+    void make_sound() const override
+    {
+        std::cout << "Wof, wof!\n";
+    }
+    void hair() const override
+    {
+        std::cout << "Fluffy\n";
+    };
+
+    std::string get_name() const override
+    {
+        return name_;
+    }
+    virtual ~Dog() = default;
+
+private:
+    std::string name_{};
+};
+
+/* Chihuahua inherits Dog
+*/
+class Chihuahua : public Dog
+{
+public:
+    Chihuahua()
+    {
+        name_ = "Minni";
+    };
+    Chihuahua(std::string name)
+    {
+        name_ = name;
+    }
+    void make_sound() const final override
```

Comment:
Only use one of virtual, override, or final -> so these are final 🙂 
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final